src/shaders/pbr.frag

```diff
@@ -328,26 +413,65 @@ void main() {
     float n_dot_h = clamp(dot(n, halfVector), 0.0, 1.0);
     vec3 currentDiffuseContrib = vec3(0.0, 0.0, 0.0);
     vec3 currentSpecularContrib = vec3(0.0, 0.0, 0.0);
-    microfacetModel(specularReflectance,
-                    c_diff,
-                    metallic,
-                    roughness,
-                    v_dot_h,
-                    n_dot_l,
-                    n_dot_v,
-                    n_dot_h,
-                    lightRadiance,
-                    currentDiffuseContrib,
-                    currentSpecularContrib);
+    microfacetModel(specularReflectance, c_diff, metallic, roughness, v_dot_h,
+                    n_dot_l, n_dot_v, n_dot_h, lightRadiance,
+                    currentDiffuseContrib, currentSpecularContrib);
     diffuseContrib += currentDiffuseContrib;
     specularContrib += currentSpecularContrib;
   }  // for lights
 
+  diffuseContrib *= ComponentScales[DirectDiffuse];
+  specularContrib *= ComponentScales[DirectSpecular];
+
   // TODO: use ALPHA_MASK to discard fragments
   fragmentColor += vec4(diffuseContrib + specularContrib, baseColor.a);
 #endif  // if LIGHT_COUNT > 0
 
+#if defined(IMAGE_BASED_LIGHTING)
+  vec3 iblDiffuseContrib = computeIBLDiffuse(c_diff, n);
+  fragmentColor.rgb += iblDiffuseContrib;
+
+  vec3 reflection = normalize(reflect(-view, n));
+  vec3 iblSpecularContrib =
+      computeIBLSpecular(roughness, n_dot_v, specularReflectance, reflection);
+  fragmentColor.rgb += iblSpecularContrib;
+#endif  // IMAGE_BASED_LIGHTING
+
 #if defined(OBJECT_ID)
   fragmentObjectId = ObjectId;
 #endif
-}
+  // PBR equation debug
+  // "none", "Diff (l,n)", "F (l,h)", "G (l,v,h)", "D (h)", "Specular"
+  if (PbrDebugDisplay > 0) {
```

Comment:
I have a feeling that this should be hidden under a define -- even if not used, the extra shader instructions may make the code run slower than if it wouldn't be there at all.

Reply:
I understand ideally it should contain no branching in the shader. But a single `if` cause is probably OK and will not cause big difference. No?

Reply:
It was more about the extra shader instructions, not the branching. More instructions will cause the compiled shader binary to be bigger, which may result in more instruction cache, more registers / ALUs / memory accesses being needed for each execution, which then might mean the GPU can't parallelize as much. You probably remember the hard limits on shader instruction count back "in the old days", that's no longer a problem fortunately but it can still happen that for example each core now has to use a twice as big register file / instruction cache and thus only half of the cores can be run at a time.

Yes, a GPU is mostly a black box, hard to reason about, and every hardware is different, but I think it won't hurt to have this `#ifdef`'d. Of course the driver might see the opportunity on its own and recompile the shader based on how this uniform is set so you won't see any difference.

Reply:
Yeah, it makes sense to me. Thank you.